test/objectwrap_nonconstructor.cc

```diff
@@ -0,0 +1,46 @@
+#include <napi.h>
+#include "test_helper.h"
+#include <unordered_map>
+
+class NonConstructorTest :
+  public Napi::ObjectWrap<NonConstructorTest> {
+public:
+  NonConstructorTest(const Napi::CallbackInfo& info) :
+    Napi::ObjectWrap<NonConstructorTest>(info) {}
+
+  static Napi::Value NonConstructor(const Napi::CallbackInfo& info) {
+    // If called with a "true" argument, throw an exeption to test the handling.
+    if(!info[0].IsUndefined() && MaybeUnwrap(info[0].ToBoolean())) {
+      NAPI_THROW(Napi::Error::New(info.Env(), "an exception"), Napi::Value());
+    }
+    // Otherwise, act as a factory.
+    std::vector<napi_value> args;
+    for(size_t i = 0; i < info.Length(); i++) args.push_back(info[i]);
+    return MaybeUnwrap(GetConstructor(info.Env()).New(args));
+  }
+
+  static std::unordered_map<napi_env, Napi::FunctionReference*> constructors;
+
```

Comment:
I know this is leaky without an env-shutdown-hook, but it's good enough to run the tests without clobbering `Napi::Addon<T>` via `env.SetInstanceData()` as would be the typical idiom.

Reply:
Should probably add a comment to the source so nobody thinks it’s a good idea to copy paste this into a real project…

Reply:
I've added an env cleanup-hook so this test-code shouldn't leak anymore and should be adaptable to real-world projects.